### ISSUE TITLE
Why use additional multiplication of np.sqrt(attrs.shape[1]) in attribute normalization?

### ISSUE THREAD

**Issue by TrynBug** [universome/czsl#3]
Attributes Normalization (AN) in the paper is as follows:
<img src="https://latex.codecogs.com/svg.latex?a_c{\;}{\mapsto}{\;}{a_c}/{\lVert}{a_c}{\rVert}_2" />

But the code uses Attributes Normalization (AN) like this:
<img src="https://latex.codecogs.com/svg.latex?a_c{\;}{\mapsto}{\;}{a_c}/{\lVert}{a_c}{\rVert}_2{\times}{\sqrt{d_a}}" />
<img src="https://latex.codecogs.com/svg.latex?{d_a}" /> is the dimensionality of attribute vector.
And the code for attributes normalization that I found is (at the preprocessing part of the class-norm-for-czsl.ipynb file)
`attrs = attrs / attrs.norm(dim=1, keepdim=True) * np.sqrt(attrs.shape[1])`

I couldn't find about this additional multiplication in the paper.
And it seems to have a huge influence on performance.
Can you tell me why this is used?

**Comment by universome** [comment]
Hi! We added this to avoid possible technical optimization issues that might arise from dividing attributes by a too large number which would squash their values: neural networks like it when their inputs are coming from N(0, I) and doing "`a_c / ||a_c|| * sqrt(d_a)`" in this case is [conceptually similar to standardization](https://openaccess.thecvf.com/content_CVPR_2020/html/Singh_Filter_Response_Normalization_Layer_Eliminating_Batch_Dependence_in_the_Training_CVPR_2020_paper.html). Attributes normalization on its own is not supposed to work for deep models and we left it their to be more consistent with the linear case. 

I believe that (for deep models) it is possible to replace all this with simple standardization (i.e. doing `(a_c - mean(a_c)) / std(a_c`) — but I just tried this (without any other changes) and it worked _poorly_, so I suspect it is necessary to adjust the optimization hyperparams (learning rate, weight decay, scheduler, etc.) somehow. Also, there might be some other explanation but I will need to think about it.

P.S. Sorry for my late reply
